Feature/add middlewares

csc-jm opened the pull request:
### Description

Pretty standard implementation of http error middleware. Might have to adjust it later as right now it also converts all other exceptions into Internal Server Error, which might not be useful for the front-end.

### Related issues
<!-- Reference any follow up issues with "Fixes #<issue-num>." -->
Fixes #9 

### Type of change

<!-- Replace [ ] with [x] to select options. -->
<!-- Please delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)

### Changes Made

<!-- List changes made. -->
- add middleware in a separate file and reference it in server initiation

### Testing

<!-- Are any tests part of this PR. -->
<!-- Replace [ ] with [x] to select options. -->
<!-- Please delete options that are not relevant. -->

- [x] Unit Tests

otahontas commented on metadata_backend/api/middlewares.py:
```diff
@@ -0,0 +1,50 @@
+        """
+        try:
+            response = await handler(req)
+            if response.status == 404:
```
I don't think this is needed separately here, since 404's are also catched with except-clause below.

You can just `return await handler(req)`

otahontas replied:
Or we could actually add forward to frontend, where react renders 404 page?

otahontas commented on metadata_backend/api/middlewares.py:
```diff
@@ -0,0 +1,50 @@
+            return _json_exception(error.status, error)
+
+        # If something else goes wrong, convert to Internal Server Error
+        except Exception as ex:
```
Like we discussed yesterday, this might hide some bugs and lead to weird situations. If program raises exceptions instead of httpErrors, we should add try-except to where the exception was raised.

otahontas commented on metadata_backend/api/middlewares.py:
```diff
@@ -0,0 +1,50 @@
+    :returns: Response in JSON format
+    """
+    body = json.dumps({
+        'error': exception.__class__.__name__,
```
Maybe better to return error as key and reason as value? I don't think class name is important for user and correct status code is anyways always returned.

blankdots replied:
we will change this in the OpenAPI specs. so fo now I think returning the HTTP status (as redundant it is) would suffice.

otahontas commented on tests/test_middlewares.py:
```diff
@@ -0,0 +1,51 @@
+        self.assertIn("detail", await response.json())
+
+    @unittest_run_loop
+    async def test_error_middleware_does_not_affect_good_responses(self):
```
Probably no need to test this, since other tests would broke if middleware would affect them?

otahontas commented:
I added some suggestions for stuff we discussed yesterday:
- removed /{schema}, we don't need it since we have /object/{schema}
- added checks that /object/{schema} returns 404 if schema is wrong (like "stuuufyy" etc)
- fixed that AssertionError bug @csc-jm noticed, it now returns 400 for error.

@csc-jm feel free to fix these, if you'd like to handle things differently!

otahontas commented:
Btw, would be also good to log errors in middleware

blankdots commented:
yeah logging is lacking a bit at the moment, was testing yesterday the backend, but logs were quite scarce

csc-jm commented:
Added a test for the 404 case in the handlers just to see for myself how that works. Can be removed if not a necessary test. Added some logging aspect in the middleware without being completely sure, if it was the proper way. Well, it's merged now. :D

blankdots commented:
@csc-jm you can create a new PR and we can merge that as well.
